Improves generic meta-model cross validation (#2612)
openturns/openturns: python/doc/theory/meta_modeling/cross_validation.rst

.. _cross_validation:

Validation and cross validation of surrogate models
---------------------------------------------------

Introduction
~~~~~~~~~~~~

In ordinary validation of a surrogate model, we hold some observations from the
sample and train the model on the remaining observations.
Then we test the surrogate model against the hold out observations.
One of the problems of this method is that the surrogate
model is trained using only a subset of the available data.

> As a consequence, the estimated error can be pessimistic.

Author: ```suggestion
As a consequence, the estimated error can be pessimistic or, on the contrary, overly optimistic.
```

> In cross-validation, the roles of the train and test samples are exchanged.
> Hence, more of the available data is used, resulting in an improved
> accuracy of the estimated error of the metamodel.

Author: ```suggestion
In cross-validation, the entirety of the data set is sequentially used for either training or validation.
Hence, more of the available data is used, resulting in an improved
accuracy of the estimated error of the metamodel.
```

> accuracy of the estimated error of the metamodel.

Author: You are right: alternating between both words is confusing.

But "surrogate model" cannot be specific to the French community... because it is not French! So, let's benchmark this.
- "Metamodel": 14 languages
- "Surrogate model": 3 languages

So "Metamodel" wins.

![image](https://github.com/openturns/openturns/assets/31351465/adc9af62-37ce-44d0-8ce0-846e5712ecd3)

![image](https://github.com/openturns/openturns/assets/31351465/4d522760-d2d2-49b1-bb20-98ff80c7c450)

> accuracy of the estimated error of the metamodel.

Author: Fixed at a7221cdfd8c7e9842c53ef8bfbd6f5d70791831a

> accuracy of the estimated error of the metamodel.

Reviewer: > You are right: alternating between both words is confusing.
> 
> But "surrogate model" cannot be specific to the French community... because it is not French! So, let's benchmark this.
> 
>     * "Metamodel": 14 languages
> 
>     * "Surrogate model": 3 languages
> 
> 
> So "Metamodel" wins.
> 
> ![image](https://private-user-images.githubusercontent.com/31351465/326035352-adc9af62-37ce-44d0-8ce0-846e5712ecd3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQzMzMzMjksIm5iZiI6MTcxNDMzMzAyOSwicGF0aCI6Ii8zMTM1MTQ2NS8zMjYwMzUzNTItYWRjOWFmNjItMzdjZS00NGQwLThjZTAtODQ2ZTU3MTJlY2QzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDI4VDE5MzcwOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTczMzlkNmM1MDU0M2QxMzc1YWI2MTczNWQ1MDBlZTA0YzFmM2ZlY2ZiYTVhNzQ5N2NhMmRjMTViMzAxYjZlMTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.WgcG8LIZYMJjNTAALQyO07Vegco4sfvmZVwAXKamRVs)
> 
> ![image](https://private-user-images.githubusercontent.com/31351465/326035469-4d522760-d2d2-49b1-bb20-98ff80c7c450.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQzMzMzMjksIm5iZiI6MTcxNDMzMzAyOSwicGF0aCI6Ii8zMTM1MTQ2NS8zMjYwMzU0NjktNGQ1MjI3NjAtZDJkMi00OWIxLWJiMjAtOThmZjgwYzdjNDUwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDI4VDE5MzcwOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY5ZjgzMGViZTA1MGNlNWYwMDljZGM4YzNiZjlmZGQ5MmFjNGFlMmQ1ZTU5OTY0OTVjODU0Yzg1OWRhOTdkYTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.yXyeDv6Qhkk9m8nAQQ-L-coEiiYRJYVnVJlv1zlCB8Y)

Actually, the Wikipedia article about "metamodeling" is not really about what our community calls a "metamodel"/"surrogate model" but seems to be rather a software engineering concept. The intro of the wikipedia page says:

_Not to be confused with [Surrogate model](https://en.wikipedia.org/wiki/Surrogate_model)._

Things are a little muddled by the "Type of metamodels" section which lists Kriging and neural networks... as software engineering concepts!

Since you mentioned the 14 languages of "Metamodeling" I had a look at the French version. You can see stuff like:

*Dans le cadre d'une [modélisation](https://fr.wikipedia.org/wiki/Mod%C3%A9lisation), un métamodèle est un [modèle](https://fr.wikipedia.org/wiki/Mod%C3%A8le_(informatique)) dont les instances sont elles-mêmes des modèles.*

*Quatre aspects sont nécessaires à la spécification d'un métamodèle : une [syntaxe abstraite](https://fr.wikipedia.org/wiki/Syntaxe_abstraite), une syntaxe concrète, une sémantique statique et une sémantique dynamique[1](https://fr.wikipedia.org/wiki/M%C3%A9tamod%C3%A8le#cite_note-:0-1),[4](https://fr.wikipedia.org/wiki/M%C3%A9tamod%C3%A8le#cite_note-:1-4). Idéalement, ces éléments sont définis à l'aide d'un langage formel[4](https://fr.wikipedia.org/wiki/M%C3%A9tamod%C3%A8le#cite_note-:1-4). Pour cette raison, un métamodèle est aussi parfois défini comme le modèle d'un [langage de modélisation](https://fr.wikipedia.org/wiki/Langage_de_mod%C3%A9lisation)[5](https://fr.wikipedia.org/wiki/M%C3%A9tamod%C3%A8le#cite_note-5).*

On the contrary, when you look at the "surrogate model" page (which has no French equivalent) the intro is:

*A surrogate model is an engineering method used when an outcome of interest cannot be easily measured or computed, so an approximate [mathematical model](https://en.wikipedia.org/wiki/Mathematical_model) of the outcome is used instead.*

My conclusion is that Wikipedia rather supports using "surrogate model" everywhere.

Author: Assuming you are right (and it is likely to be so), then we have an issue here, because the word "metamodel" is used throughout the library. For example, the class is named "MetaModelValidation". This is one of the reasons I used this word. If surrogate model is a best choice, then "SurrogateModelValidation" would be best.

Reviewer: Yes, this is clearly beyond the scope of this PR, I only posted this for future reference.

One of the problems of cross-validation is that we have to train
the surrogate model several times which can be time consuming.
In this case, the fast methods that we present below can be useful.

Let :math:`\model: \set{X} \rightarrow \Rset` be a physical
model where :math:`\set{X} \subseteq \Rset^{n_x}`
is the domain of the input parameters :math:`\vect{x}`.
Let :math:`\metaModel` be a surrogate model of :math:`\model`, i.e.
an approximation of the function.
Once the metamodel :math:`\metaModel`
of the original numerical model :math:`\model` has been

> built up, we may estimate the mean squared error, i.e. the

Author: I do not understand this suggestion, since the "built" verb is already used in this sentence. Can you provide a more specific suggestion?

> built up, we may estimate the mean squared error, i.e. the

Reviewer: I think Julien would like to replace "built up" with "built"

> built up, we may estimate the mean squared error, i.e. the

Author: ```suggestion
built, we may estimate the mean squared error, i.e. the
```

> built up, we may estimate the mean squared error, i.e. the

Author: ```suggestion
built, we may estimate the mean squared error, i.e. the
```

The fraction of variance unexplained is a relative mean squared error.
The coefficient of determination is:

.. math::

    R^2 & = 1 - \operatorname{FVU} \\
    & = 1 - \frac{\operatorname{MSE}\left(\metaModel\right)}{\Var{Y}}

Simple validation
~~~~~~~~~~~~~~~~~

In the ordinary or naive validation method, we divide the data sample (i.e.
the experimental design) into two independent sub-samples:
the training and the test set.
In this case, the metamodel :math:`\metaModel`
is built from one sub-sample, i.e. the *training set*, and its
performance is assessed by comparing its predictions to the other
subset, i.e. the *test set*.
A single split will lead to a validation estimate.
When several splits are created, the cross-validation error
estimate is obtained by averaging over the splits.
When the coefficient of determination is evaluated on a test set which is
independent from the training set, the coefficient of determination is
called the :math:`Q^2` score.

.. figure:: ../../_static/train_test_cross_validation.png
    :align: center
    :alt: Validation of a surrogate
    :width: 50%

    *Figure 1.* Validation using one single split.

> Let :math:`n \in \Nset` be an integer representing the sample size.

Author: ```suggestion
Let :math:`\sampleSize \in \Nset` be an integer representing the sample size.
```

    = \frac{1}{n} \sum_{j = 1}^n \left( y^{(j)} - \metaModel \left(\vect{x}^{(j)}\right) \right)^2

The previous equation can be equivalently expressed depending on
the model since :math:`y^{(j)} = \model\left(\vect{x}^{(j)}\right)`.
It seems, however, more consistent to use :math:`y^{(j)}` because the
true model :math:`g` is unknown (otherwise we would not use a
surrogate).

The sample relative mean squared error is:

  .. math::

      \widehat{\varepsilon}_{MSE}\left(\metaModel\right)
      = \frac{\widehat{\operatorname{MSE}}\left(\metaModel\right)}{\widehat{\sigma}^2(Y)}

where :math:`\widehat{\sigma}^2(Y)` is the sample variance of the random output:

.. math::

      \widehat{\sigma}^2(Y)
      = \frac{1}{n - 1} \sum_{j = 1}^n \left( y^{(j)} -   \bar{y} \right)^2

where :math:`\bar{y}` is the sample mean of the output:

.. math::

    \bar{y} = \frac{1}{n} \sum_{j = 1}^n y^{(j)}.

If the test set :math:`\set{D}` is not independent from the training set
(the set used to calibrate the metamodel), then the previous estimator
may underestimate the true value of the mean squared error.
In order to create a test set independent from the training set, a

> simple method is to split the data set into two parts.

Author: Assuming that the sample is i.i.d., then splitting the data into non overlapping sub-samples guarantees that the two sub-samples are independent. When you suggest that splitting is more robust, what do you compare to? That is to say, splitting is, according to you, more robust than what other method?

The drawback of this method is that this reduces the size of the training

> set, so that the mean squared error evaluated on the test set can be pessimistic.

Author: The particular sentence that you point means that splitting the data into two non overlapping sub-samples reduces the amount of data available for both tasks:
- The sub-sample size used for training is smaller. This implies that the accuracy of the metamodel is lower than the accuracy of the metamodel trained on the full data, which is what we are interested in. Hence, the coefficients estimated from the sub-sample are less accurately estimated. Hence, the R2 score is most of the times pessimistic.
- The sub-sample size used for testing is smaller. This implies that the estimated R2 score is less accurately estimated. This reduced-quality estimate of the R2 score can be optimistic or pessimistic, depending if it is larger or smaller than the true value. Whatever the order, its accuracy is smaller than if we used all the available data.

My conclusion is that the current sentence is correct.

> set, so that the mean squared error evaluated on the test set can be pessimistic.

Author: ```suggestion
set, so that the mean squared error evaluated on the test set can be pessimistic (because the 
metamodel is trained with less data) or optimistic (because the R2 score has a greater variability).
```

> In order to create a test set independent from the training set, a
> simple method is to split the data set into two parts.

Author: ```suggestion
Assuming that the sample is made of i.i.d. observations, 
in order to create a test set independent from the training set, a
simple method is to split the data set into two parts.
```

The corresponding set of observation indices is:

.. math::
    \set{S}_{-j} = \{1, ..., j - 1, j + 1, ..., n\},

the corresponding input observations are:

.. math::
    \set{D} \setminus \{\vect{x}^{(j)}\}
    = \left\{\vect{x}^{(1)}, ..., \vect{x}^{(j - 1)},
    \vect{x}^{(j + 1)}, ..., \vect{x}^{(n)}\right\}
    = \left\{\vect{x}^{(j)}, \; j \in \set{S}_{-j}\right\}

and the corresponding output observations are:

.. math::
    \left\{y^{(1)}, ..., y^{(j - 1)}, y^{(j + 1)}, ..., y^{(n)}\right\}.

The leave-one-out residual is defined as the difference between the model evaluation at
:math:`\vect{x}^{(j)}` and its leave-one-out prediction (see [blatman2009]_
eq. 4.26 page 85):

.. math::

    \Delta^{(j)}
    = y^{(j)} - \metaModel^{(-j)}\left(\vect{x}^{(j)}\right)

We repeat this process for all observations in the experimental
design and obtain the predicted residuals
:math:`\Delta^{(j)}` for :math:`j = 1, \dots, n`.
Finally, the LOO mean squared error estimator is:

.. math::

    \widehat{\operatorname{MSE}}_{LOO}
    & =  \frac{1}{n} \sum_{j = 1}^n \left( \Delta^{(j)} \right)^2

One of the drawbacks of the naive method is that it may require

> to estimate :math:`n` different surrogate models.

Author: I agree. But the purpose of this specific sentence is to introduce the next section, which is the fast LOO CV. This is why I would rather leave it here.

> If :math:`n` is large or if training each surrogate model is costly,

Author: ```suggestion
If :math:`\sampleSize` is large or if training each surrogate model is costly,
```

Assume that the number of folds divides the sample size.
Mathematically, this means that :math:`k` divides :math:`n`.
In this special case, each fold has the same number of observations:

.. math::

    n_\ell = \frac{n}{k}

for :math:`\ell = 1, ..., k`. Hence all local K-Fold MSE have the
same weight and we have :math:`\frac{n_\ell}{n} = \frac{1}{k}`
for :math:`\ell = 1, ..., k`.
This implies that the K-Fold mean squared error has a  particularly simple expression
([deisenroth2020]_ eq. 8.13 page 264):

.. math::
  :label: kfoldMeanEqual

   \widehat{\operatorname{MSE}}_{KFold}
   = \frac{1}{k} \sum_{\ell = 1}^{k} \widehat{\operatorname{MSE}}^{(\ell)}.

Fast K-Fold cross-validation of a linear model
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

In this section, we present a fast version of the K-Fold cross-validation
that can be used for a linear model.
While evaluating the mean squared error with the fast LOO formula involves the division by :math:`1 - h_{jj}`,
using the fast K-Fold method involves the resolution of a linear system of equations.
The equations introduced by [shao1993]_ are presented in [suzuki2020]_ (proposition 14 page 71).

> For any :math:`\ell \in \{1, ..., k\}`, let :math:`\boldsymbol{D}_{\ell} \in \Rset^{n_\ell \times m}`

Author: You are right: this is nx+1.
```suggestion
For any :math:`\ell \in \{1, ..., k\}`, let :math:`\boldsymbol{D}_{\ell} \in \Rset^{n_\ell \times (n_x + 1)}`
```
